fs/span-manager/span_manager.go

```diff
@@ -162,6 +163,49 @@ func (m *SpanManager) buildAllSpans() {
 	}
 }
 
+// FetchSingleSpan invokes the reader to fetch the span in the background and cache it.
+// It is invoked by the BackgroundFetcher.
+func (m *SpanManager) FetchSingleSpan(spanID soci.SpanID) error {
+	if spanID > m.ztoc.CompressionInfo.MaxSpanID {
+		return ErrExceedMaxSpan
+	}
+
+	s := m.spans[spanID]
+	s.mu.Lock()
+	state := s.state.Load().(spanState)
+
+	// Only fetch if the span hasn't been requested yet.
+	if state != unrequested {
+		s.mu.Unlock()
+		return ErrSpanStateNotUnrequested
+	}
+
+	s.setState(requested)
+	s.mu.Unlock()
+
+	compressedSize := s.endCompOffset - s.startCompOffset
+	compressedBuf := make([]byte, compressedSize)
+
+	n, err := m.r.ReadAt(compressedBuf, int64(s.startCompOffset))
+	if err != nil {
+		return err
+	}
+
+	if n != len(compressedBuf) {
+		return fmt.Errorf("unexpected data size for reading compressed span. read = %d, expected = %d", n, len(compressedBuf))
+	}
+
+	if err := m.verifySpanContents(compressedBuf, spanID); err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	m.addSpanToCache(fmt.Sprintf("%d", spanID), compressedBuf)
```

Comment:
We should verify that state is still `requested` before doing this or else we have a race where we could replace previously uncompressed data with compressed data.